### PR TITLE
Added StringLiteralDuplication rule

### PR DESCRIPTION
**Pull request by schalkms**

### CODE REVIEW

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/StringLiteralDuplication.kt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+	internal inner class StringLiteralVisitor : DetektVisitor() {
+
+		var size = 0
+
+		override fun visitLiteralStringTemplateEntry(entry: KtLiteralStringTemplateEntry) {
+			increment()
+		}
+
+		private fun increment() {
```
Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/StringLiteralDuplication.kt]: Why not inline this function?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/StringLiteralDuplication.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+			size++
+		}
+	}
+}
+
+private const val DEFAULT_DUPLICATION = 2
```
Comment (vanniktech) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/complexity/StringLiteralDuplication.kt]: I find 2 quite low as a default.

**Comment by schalkms** [comment]
@vanniktech 
The default value is of course discussable.
I think the string duplication should also exclude whitespaces and common strings like 
- "."
- ","
I am waiting for feedback. I am also willing to fix the following duplicates after giving feedback.

```
StringLiteralDuplication - 5/2 - [CodeSmell.kt] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/CodeSmell.kt:1:1
	
StringLiteralDuplication - 3/2 - [Location.kt] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Location.kt:1:1
	
StringLiteralDuplication - 3/2 - [Signatures.kt] at detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Signatures.kt:1:1
	
StringLiteralDuplication - 3/2 - [SpacingAfterKeyword.kt] at detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/SpacingAfterKeyword.kt:1:1
	
StringLiteralDuplication - 4/2 - [SpacingAroundBraces.kt] at detekt-formatting/src/main/kotlin/io/gitlab/arturbosch/detekt/formatting/SpacingAroundBraces.kt:1:1
	
StringLiteralDuplication - 7/2 - [DetektPlugin.kt] at detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/DetektPlugin.kt:1:1
	
StringLiteralDuplication - 3/2 - [NamingConventionViolation.kt] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NamingConventionViolation.kt:1:1
	
StringLiteralDuplication - 13/2 - [NamingConventionViolation.kt] at detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/style/NamingConventionViolation.kt:1:1
```

**Comment by vanniktech** [comment]
How about passing a list of values that should be ignored and by default default to . and ,

**Comment by arturbosch** [comment]
An ignore list sounds good to me. 
I like the rule! Maybe a name like `ConsiderConstantString` would suit it better?

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/ComplexityProvider.kt [hunk 3]
```diff
@@ -25,6 +26,7 @@ class ComplexityProvider : RuleSetProvider {
 				LongMethod(config),
 				LargeClass(config),
 				ComplexMethod(config),
+				ConsiderConstantString(config),
```
Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/ComplexityProvider.kt]: Sorry my bad. ConsiderStringConstant is a better name. 

Why ComplexityProvider? I think style suits it better?

**Comment by arturbosch** [comment]
I like the configuration options!

I will wait some more days so we can figure out the threshold value. 2 seems to be too low, but also reasonable ...

File: detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/ComplexityProvider.kt (hunk 3 above)
Comment (schalkms) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/ComplexityProvider.kt]: I disagree. It is basically code duplication. I think multiple occurrences of hard coded strings are making maintenance more difficult.

Comment (arturbosch) [inline detekt-rules/src/main/kotlin/io/gitlab/arturbosch/detekt/rules/providers/ComplexityProvider.kt]: Hmm now I'm thinking about that the name StringLiteralDuplication was good though :/ and the description should suggest using a constant.

**Comment by arturbosch** [comment]
i think some of the `StringLiteralDuplication` findings you posted above are false positive. I do not see any duplicated strings in CodeSmell, Location, NamingConvention?

**Comment by schalkms** [comment]
@arturbosch This is an outdated version.

**Comment by schalkms** [comment]
I also did some research concerning this rule. SonarJava also this rule. They have a threshold of 3. However, this threshold comparing it to detekts means setting the threshold for this rule to 2.
IMHO the threshold for detekt rules is confusing and thus should be refactored.

Furthermore, SonarJava also doesn't detect strings with less than 5 characters.

reference:
https://www.sonarsource.com/products/codeanalyzers/sonarjava/rules.html#RSPEC-1192

What do you think? @arturbosch @vanniktech @Mauin

**Comment by arturbosch** [comment]
I think we should follow the sonarqube way and exclude strings with size < 5 and have a threshold of 3 (2 in detekt).

Seems like that sonar uses value >= threshold. Detekt (well I introduced it) uses value > threshold. So 2 seems low, effectlively it is 3. I will open a issue to discuss the which threshold definition we should use.

**Comment by arturbosch** [comment]
:+1:

**Pull request merged**